### PR TITLE
Native + Lucene Fusion index as default

### PR DESCRIPTION
**Pull request by burqen**
## Fusion index as default
Fusion index = a fusion between native and Lucene index.
Native index is built on top of GB+Tree and can only handle numbers.
Lucene index is old Lucene implementation.

Fusion index works by wrapping Native and Lucene into one logical index where all numbers are directed to Native and the rest (including all arrays and all composite keys) to Lucene.

First benchmarks show a write performance improvement of between 1.5-4x.

## Default behaviour
- All newly created indexes will be 'fusion indexes'.
- Old indexes (created with only Lucene) will not be migrated but continue as 'Lucene only' indexes. Drop and recreate to use fusion index. This is recommended for indexes on number focused properties (not composite or arrays). For indexes over Strings there is no use to move to fusion.

## Migration
Migrating from <3.3 will keep all indexes around and will work just fine as Lucene only. Drop and recreate to move to fusion index.

## Internal config to disable
An internal unsupported config option is introduced to disable fusion index as default. In this case, Lucene will be used instead. Disabling fusion index will NOT migrate already created fusion indexes. Those will have to be dropped and recreated to use the new default (Lucene).

## TL;DR
- New fusion index = native + lucene.
- Fusion is default.
- When creating new index, default is used.
- No migration is done for already existing indexes. Can be migrated to default by drop and recreate.
- Turn fusion index of with internal unsupported config option.
- Faster, Harder, Scooter.

co-author: @tinwelint 

[cl] Introduce new native index for numbers.

### CODE REVIEW

File: community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        public void add( CleanupJob job )
+        {
+            if ( job.needed() )
+            {
+                throw new IllegalStateException( "Consistency checker should not do recovery" );
+            }
+        }
+    }
+
+    public static final RecoveryCleanupWorkCollector RECOVERY_PREVENTING_COLLECTOR = new RecoveryPreventingCollector();
```
Comment (MishaDemianenko) [inline community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java]: Please fields first, public member and them private member

Comment (MishaDemianenko) [inline community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java]: not sure what collectors class and definition does in SchemaIndexExtensionLoader, even if it's passed as one of the parameters to instantiate extensions

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import static org.neo4j.kernel.extension.KernelExtensionUtil.servicesClassPathEntryInformation;
+
+/**
+ * Selects the candidate with highest priority (assumed to implement {@link Comparable}) and returns
+ * in {@link #select(Class, Iterable)}, but keeps the others for access too.
+ */
+public class AllByPrioritySelectionStrategy<T> implements DependencyResolver.SelectionStrategy
+{
+    @SuppressWarnings( "rawtypes" )
+    private List lowerPrioritizedCandidates = Collections.emptyList();
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java]: `private List<T> lowerPrioritizedCandidates = Collections.emptyList();` then you do not need to suppress warnings

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import org.neo4j.graphdb.DependencyResolver;
+import org.neo4j.helpers.collection.Iterables;
+
+import static org.neo4j.kernel.extension.KernelExtensionUtil.servicesClassPathEntryInformation;
+
+/**
+ * Selects the candidate with highest priority (assumed to implement {@link Comparable}) and returns
+ * in {@link #select(Class, Iterable)}, but keeps the others for access too.
+ */
+public class AllByPrioritySelectionStrategy<T> implements DependencyResolver.SelectionStrategy
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java]: `<T extends Comparable>` instead `<T>`

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            throw new IllegalArgumentException( "Could not resolve dependency of type: " +
+                                                type.getName() + ". " + servicesClassPathEntryInformation() );
+        }
+        Collections.sort( all, Collections.reverseOrder() );
+        R highest = (R) all.remove( 0 );
+        lowerPrioritizedCandidates = all;
+        return highest;
+    }
+
+    @SuppressWarnings( "unchecked" )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java]: no need to suppress warnings

File: community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+public class AllByPrioritySelectionStrategy<T> implements DependencyResolver.SelectionStrategy
+{
+    @SuppressWarnings( "rawtypes" )
+    private List lowerPrioritizedCandidates = Collections.emptyList();
+
+    @SuppressWarnings( {"rawtypes", "unchecked"} )
+    @Override
+    public <R> R select( Class<R> type, Iterable<R> candidates ) throws IllegalArgumentException
+    {
+        List<Comparable> all = (List<Comparable>) Iterables.asList( candidates );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/extension/dependency/AllByPrioritySelectionStrategy.java]: ` List<T> all = (List<T>) Iterables.asList( candidates );` instead

File: community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java [hunk 6]
```diff
@@ -556,11 +556,13 @@ public void shouldReportNodesThatAreNotIndexed() throws Exception
...
             IndexRule indexRule = indexRuleIterator.next();
-            IndexAccessor accessor = fixture.directStoreAccess().indexes().getOnlineAccessor(
-                    indexRule.getId(), indexRule.getIndexDescriptor(), samplingConfig );
+            IndexAccessor accessor = fixture.directStoreAccess().indexes().
+                    apply( indexRule.getProviderDescriptor() ).getOnlineAccessor(
+                            indexRule.getId(), indexRule.getIndexDescriptor(), samplingConfig );
             IndexUpdater updater = accessor.newUpdater( IndexUpdateMode.ONLINE );
             updater.remove( asPrimitiveLongSet( indexedNodes ) );
             updater.close();
+            accessor.force();
```
Comment (MishaDemianenko) [inline community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java:565]: what's the point doing force specifically before closing accessor?

File: community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java [hunk 7]
```diff
@@ -387,9 +387,11 @@ private int search( PageCursor cursor, KEY key, KEY readKey, int keyCount )
     private boolean cursorIsAtExpectedLocation( PageCursor cursor )
     {
         assert currentLevel >= 0 : "Uninitialized tree logic, currentLevel:" + currentLevel;
-        assert cursor.getCurrentPageId() == levels[currentLevel].treeNodeId : "Expected cursor to be at page:" +
-                    levels[currentLevel].treeNodeId + " at level:" + currentLevel + ", but was at page:" +
-                    cursor.getCurrentPageId();
+        long currentPageId = cursor.getCurrentPageId();
+        long expectedPageId = levels[currentLevel].treeNodeId;
+        assert currentPageId == expectedPageId : "Expected cursor to be at page:" +
```
Comment (MishaDemianenko) [inline community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java:392]: noo, that was one of my favourite non-readable statements 😞

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java [hunk 8]
```diff
@@ -394,6 +394,10 @@
     public static final Setting<Boolean> multi_threaded_schema_index_population_enabled =
             setting( "unsupported.dbms.multi_threaded_schema_index_population_enabled", BOOLEAN, TRUE );
 
+    @Internal
+    public static final Setting<Boolean> enable_native_schema_index =
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:404]: surprising option

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java [hunk 9]
```diff
@@ -181,23 +160,10 @@ private void awaitCompletion()
      */
     private void batchUpdate( IndexPopulation population, IndexEntryUpdate<?> update )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java]: this method is not used at all now

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java [hunk 10]
```diff
@@ -206,18 +172,7 @@ private void flushIfNeeded( IndexPopulation population, List<IndexEntryUpdate<?>
...
-            Map.Entry<IndexPopulation,List<IndexEntryUpdate<?>>> entry = entries.next();
-            IndexPopulation population = entry.getKey();
-            List<IndexEntryUpdate<?>> updates = entry.getValue();
-            entries.remove();
-            if ( updates != null && !updates.isEmpty() )
-            {
-                flush( population, updates );
-            }
-        }
+        populations.forEach( population -> flush( population, population.takeCurrentBatch() ) );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/BatchingMultipleIndexPopulator.java]: there is no need to pass second param here since population is already passed as first one

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java [hunk 11]
```diff
@@ -86,6 +89,7 @@
 
     public static final String QUEUE_THRESHOLD_NAME = "queue_threshold";
     private final int QUEUE_THRESHOLD = FeatureToggles.getInteger( getClass(), QUEUE_THRESHOLD_NAME, 20_000 );
+    private final int BATCH_SIZE = FeatureToggles.getInteger( BatchingMultipleIndexPopulator.class, BATCH_SIZE_NAME, 10_000 );
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java]: feature toggle is moved but class is the same? why?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java [hunk 12]
```diff
@@ -80,29 +78,46 @@ public IndexAccessor getOnlineAccessor( long indexId, IndexDescriptor descriptor
     @Override
     public String getPopulationFailure( long indexId ) throws IllegalStateException
     {
-        String nativeFailure = nativeProvider.getPopulationFailure( indexId );
-        String luceneFailure = luceneProvider.getPopulationFailure( indexId );
-        if ( nativeFailure != null )
+        String nativeFailure = null;
+        try
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:82]: what is the sense in throwing and catching useless exceptions instead of updating providers to not throw exceptions if not failed?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java [hunk 13]
```diff
@@ -80,29 +78,46 @@ public IndexAccessor getOnlineAccessor( long indexId, IndexDescriptor descriptor
...
         {
-            return luceneFailure;
+            luceneFailure = luceneProvider.getPopulationFailure( indexId );
+        }
+        catch ( IllegalStateException e )
+        {   // Just catch
         }
+
+        if ( nativeFailure != null || luceneFailure != null )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:98]: please use `StringUtils.isNotEmpty` instead

File: community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java [hunk 14]
```diff
@@ -408,9 +410,16 @@ public void start() throws IOException
...
         life.add( recoveryCleanupWorkCollector );
 
-        schemaIndexProvider = dependencyResolver.resolveDependency( SchemaIndexProvider.class,
-                HighestSelectionStrategy.getInstance() );
-        dependencies.satisfyDependency( schemaIndexProvider );
+        AllByPrioritySelectionStrategy<SchemaIndexProvider> indexProviderSelection = new AllByPrioritySelectionStrategy<>();
+        SchemaIndexProvider defaultIndexProvider =
+                dependencyResolver.resolveDependency( SchemaIndexProvider.class, indexProviderSelection );
+
+        // todo do we really need to have this dependency?
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:416]: TODO

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java [hunk 15]
```diff
@@ -1,199 +0,0 @@
-/*
```
Comment (MishaDemianenko) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java:1]: Not quite sure why this test was removed, we do not cancel sampling anymore?

**Review by MishaDemianenko — CHANGES_REQUESTED** [review]
Couple of small comments here and there

File: community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java (hunk 1 above)
Comment (tinwelint) [reply community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java]: Well, this class is a utility used in a bunch of places in consistency checker. It loads schema index provider kernel extensions and consistency checker is not allowed to do recovery. Where should it go?

File: community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java (hunk 6 above)
Comment (tinwelint) [reply community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java:565]: That's the general contract and something that native indexing hooks in on to checkpoint properly. It's just that lucene happened to make do without it.

File: community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java (hunk 7 above)
Comment (tinwelint) [reply community/index/src/main/java/org/neo4j/index/internal/gbptree/InternalTreeLogic.java:392]: Bugger, well I'd guess it wasn't the last of its kind :)

File: community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java (hunk 8 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/graphdb/factory/GraphDatabaseSettings.java:404]: We went for it since it's easier to manage, also in tests. Now that we've gotten all fixed and working we can investigate if it would be tricky to manage this if it were a toggle instead. Hang on.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java (hunk 11 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java]: Something of a confusion move at a time we were confused about multi index populator. Forgot about it... we can move it back.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java (hunk 13 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:98]: I think we want to check `null` explicitly. Empty string means something else in this case.

File: community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java (hunk 14 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:416]: Ah, to be investigated...

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java (hunk 15 above)
Comment (tinwelint) [reply community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexSamplingCancellationTest.java:1]: It only tested that the in-memory test index `HashBasedIndex` sample method threw exception in a particular case. I think it was mis-written to begin with.

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java (hunk 12 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSchemaIndexProvider.java:82]: `getPopulationFailure` is designed to throw `IllegalStateException` if called for index that isn't in `FAILED` state, sensible methinks.

File: community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java (hunk 14 above)
Comment (tinwelint) [reply community/kernel/src/main/java/org/neo4j/kernel/NeoStoreDataSource.java:416]: It was left there because that's how it was previously. Now the `SchemaIndexProviderMap` is used instead. It's not strictly required for us to remove this right now, since it may require some time and testing. We've created a card for it now.

**Comment by tinwelint** [comment]
I think we need to look at `SchemaIndexHaIT` dude, it fails consisently on this branch... not very often at all otherwise. I think it's simply a kernel extension loading thingie.

File: community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java (hunk 1 above)
Comment (MishaDemianenko) [reply community/consistency-check/src/main/java/org/neo4j/consistency/internal/SchemaIndexExtensionLoader.java]: well if you will need in any other test ever RecoveryPreventingCollector I would say that SchemaIndexExtensionLoader would be the last place where i would expect to find it, even in separate class is better imho

File: community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java (hunk 6 above)
Comment (MishaDemianenko) [reply community/consistency-check/src/test/java/org/neo4j/consistency/checking/full/FullCheckIntegrationTest.java:565]: I mean if you close it without force it should do that automatically right?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java (hunk 11 above)
Comment (MishaDemianenko) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/MultipleIndexPopulator.java]: The option still present in this class

**Comment by MishaDemianenko** [comment]
@tinwelint @burqen there is one small thing still please fix that otherwise its good and approved

**Review by MishaDemianenko — APPROVED** [review]

**Comment by tinwelint** [comment]
The build passed, although the Github <--> CI integration didn't quite get that bit. @burqen was there some issue with leaving open files behind (as shown in some benchmark)? If so perhaps we should hold off merge until we know more.

**Comment by burqen** [comment]
@tinwelint It's a bit unclear. There does not seem to be any leaking files but we need some more investigation before merging.
